plugins/gimp/file-jxl-save.cc

```diff
@@ -5,128 +5,884 @@
-    io.metadata.m.SetUintSamples(bits_per_sample);
-  }
-
-  const GeglRectangle rect = *gegl_buffer_get_extent(gegl_buffer);
-  std::vector<float> pixel_data(rect.width * rect.height * (3 + has_alpha));
-  gegl_buffer_get(gegl_buffer, &rect, 1., format, pixel_data.data(),
-                  GEGL_AUTO_ROWSTRIDE, GEGL_ABYSS_NONE);
-  g_clear_object(&gegl_buffer);
-
-  JXL_RETURN_IF_ERROR(
-      read_buffer(rect.width, rect.height, pixel_data, std::move(icc), &io));
-  CompressParams params;
-  PassesEncoderState encoder_state;
-  PaddedBytes compressed;
-  ThreadPoolInternal pool;
-  params.butteraugli_distance = 1.f;
-  JXL_RETURN_IF_ERROR(EncodeFile(params, &io, &encoder_state, &compressed,
-                                 /*aux_out=*/nullptr, &pool));
-  JXL_RETURN_IF_ERROR(WriteFile(compressed, filename));
+    g_printerr("JXL Info: No ICC profile.  Exporting image anyway.\n");
+  }
+
+  gimp_save_progress.update();
+
+  jxl_save_opts.set_dimensions(gimp_image_width(image_id),
+                               gimp_image_height(image_id));
+
+  layers = gimp_image_get_layers(image_id, &nlayers);
+
+  for (gint32 i = 0; i < nlayers; i++) {
+    if (gimp_drawable_has_alpha(layers[i])) {
+      jxl_save_opts.has_alpha = true;
+      break;
+    }
+  }
+
+  gimp_save_progress.update();
+
+  // JxlEncoderAddImageFrame() doesn't currently support
+  // JXL_TYPE_UINT32.  Rebuilding babl_format to match
+  // the JxlPixelFormat to allow  export to UINT16.
+  //
+  // When this is no longer necessary, will be able to
+  // replace extraneous code with just:
+  //    native_format = gegl_buffer_get_format(buffer);
+  //
+  const Babl* native_format;
+  jxl_save_opts.set_precision(gimp_image_get_precision(image_id));
+  jxl_save_opts.set_model(gimp_drawable_type(drawable_id));
+  native_format = babl_format(jxl_save_opts.babl_format_str.c_str());
+
+  gimp_save_progress.update();
+
+  // multi-threaded parallel runner.
+  auto runner = JxlResizableParallelRunnerMake(nullptr);
+
+  JxlResizableParallelRunnerSetThreads(
+      runner.get(),
+      JxlResizableParallelRunnerSuggestThreads(jxl_save_opts.basic_info.xsize,
+                                               jxl_save_opts.basic_info.ysize));
+
+  auto enc = JxlEncoderMake(/*memory_manager=*/nullptr);
+  JxlEncoderUseContainer(enc.get(), jxl_save_opts.use_container);
+
+  if (JXL_ENC_SUCCESS != JxlEncoderSetParallelRunner(enc.get(),
+                                                     JxlResizableParallelRunner,
+                                                     runner.get())) {
+    g_printerr("JXL Error: JxlEncoderSetParallelRunner failed\n");
+    return false;
+  }
+
+  if (JXL_ENC_SUCCESS !=
+      JxlEncoderSetBasicInfo(enc.get(), &jxl_save_opts.basic_info)) {
+    g_printerr("JXL Error: JxlEncoderSetBasicInfo failed\n");
+    return false;
+  }
+
+  // try to use ICC profile
+  if (icc.size() > 0 && !gimp_color_profile_is_gray) {
```

Comment:
`gimp_color_profile_is_gray` is using the address of the function as a boolean, instead of calling it. I'll send a fix for this.

Reply:
Thanks.  I saw that after it had already been merged.  It was supposed to be `!jxl_save_opts.is_gray`.